ex3/k_furuta/main.py

```diff
@@ -0,0 +1,222 @@
+import numpy as np
+from sklearn.utils.extmath import cartesian
+
+
+def extend_data(data, dim):
+    """コマンドライン引数で指定した累乗で拡張した特徴量を生成する関数.
+
+    Parameters
+    -------------
+    data : np.ndarray, shape=(the number of sample, the kinds of data)
+        特徴量の二次元配列
+    dim : int
+        各変数の次元数([1,2]ならば1,x^1,y^1,y^2)
+    Returns
+    ------------
+    extended_data : ndarray, shape=(the number of sample, 1 + sum(dim))
+        拡張されたデータ
+    """
+    # 切片用の配列を追加
+    extended_data = np.ones((data.shape[0], 1))
+
+    for i in range(len(dim)):
+        for j in range(dim[i]):
+            # 指定された各次元について累乗を計算
+            extended_data = np.concatenate(
+                [extended_data, data[:, i].reshape(-1, 1) ** (j + 1)], axis=1
+            )
+
+    return extended_data
+
+
+def calc_regression(data, dim, coeff):
+    """正規方程式を解いて重みを計算する関数.
+
+    Parameters
+    -------------
+    data : np.ndarray, shape=(the number of sample, the kinds of data)
+        特徴量と推定するデータの二次元配列
+    dim : int
+        各変数の次元数([1,2]ならば1,x^1,y^1,y^2)
+    coeff : int
+        正規化の重み(0なら正規化しない)
+    Returns
+    ------------
+    weight : ndarray, shape=(1 + sum(dim))
+        回帰直線の係数
+    """
+    # dataを特徴量と推定量に分ける
+    feature = data[:, :-1]
+    target = data[:, -1]
+
+    # 指定された次数に拡張する
+    extended_feature = extend_data(feature, dim)
+
+    # 正規化行列を解いて重みを計算
+    weight = (
+        np.linalg.inv(
+            extended_feature.T @ extended_feature
+            + coeff * np.eye(extended_feature.shape[1])
+        )
+        @ extended_feature.T
+        @ target
+    )
+
+    return weight
+
+
+def parse_args():
+    """コマンドプロントから引数を受け取るための関数."""
+    parser = argparse.ArgumentParser(
+        description="Generate spectrogram and inverse transform"
+    )
+    parser.add_argument(
+        "--input-file", type=str, required=True, help="Name of input csv file"
+    )
+    parser.add_argument(
+        "--dimension",
+        type=int,
+        required=True,
+        nargs="*",
+        help="Upper limit on the order of the basis functions",
+    )
+    parser.add_argument(
+        "--number-of-point",
+        type=int,
+        default=100,
+        help="Number of points on the regression curve",
+    )
+    parser.add_argument(
+        "--coeff", type=float, default=0.1, help="Parameters of ridge regression"
+    )
+    return parser.parse_args()
+
+
+def make_polynomial_string(weight, dim):
+    """指数の情報から多項式の文字列を作成.
+
+    Parameters
+    -------------
+    weight : ndarray, shape=(1 + sum(dim))
+        回帰直線の係数
+    dim : int
+        各変数の次元数([1,2]ならば1,x^1,y^1,y^2)
+    Return
+    ------------
+    poly : str
+        多項式の文字列(latex形式)
+    """
+    index = 0
+    poly = str(round(weight[index], 2))
+    index += 1
+    for i in range(len(dim)):
+        for j in range(dim[i]):
+            poly += f"+{round(weight[index],2)}$x_{i+1}^{j+1}$"
+            index += 1
+    return poly
+
+
+def visalize(data, weight, dim, points):
```

Comment:
```suggestion
def visualize(data, weight, dim, points):
```
エディタを使っているとなかなか気づけませんが、読み手にとってスペルミスは少し気になるので注意しましょう！